enhancements/sig-architecture/82-addon-template/README.md

```diff
@@ -0,0 +1,407 @@
+// or other endpoints on hub cluster with client certificate authentication. During the addon
+// registration process, a csr will be created for each RegistrationSpec on the hub cluster. The
+// CSR can be approved automatically(Auto) or manually(None), After the csr is approved on the hub
+// cluster, the klusterlet agent will create a secret in the installNamespace for the addon agent.
+// If the RegistrationType type is KubeClient, the secret name will be "{addon name}-hub-kubeconfig"
+// whose content includes key/cert and kubeconfig. Otherwise, If the RegistrationType type is
+// CustomSigner the secret name will be "{addon name}-{signer name}-client-cert" whose content
+// includes key/cert.
+type RegistrationSpec struct {
+	// Type of the registration configuration, it supports:
+	// - KubeClient: the addon agent can access the hub kube apiserver with kube style API.
+	//   the signer name should be "kubernetes.io/kube-apiserver-client". When this type is
+	//   used, the KubeClientRegistrationConfig can be used to define the permission of the
+	//   addon agent to access the hub cluster
+	// - CustomSigner: the addon agent can access the hub cluster through user-defined endpoints.
+	//   When this type is used, the CustomSignerRegistrationConfig can be used to define how
+	//   to issue the client certificate for the addon agent.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=KubeClient;CustomSigner
+	Type RegistrationType `json:"type"`
+
+	// KubeClient holds the configuration of the KubeClient type registration
+	// +optional
+	KubeClient *KubeClientRegistrationConfig `json:"kubeClient,omitempty"`
+
+	// CustomSigner holds the configuration of the CustomSigner type registration
+	// required when the Type is CustomSigner
+	CustomSigner *CustomSignerRegistrationConfig `json:"customSigner,omitempty"`
+}
+
+type KubeClientRegistrationConfig struct {
+	// HubPermissions represent the permission configurations of the addon agent to access the hub cluster
+	// +optional
+	HubPermissions []HubPermissionConfig `json:"hubPermissions,omitempty"`
+}
+
+// HubPermissionConfig configures the permission of the addon agent to access the hub cluster.
+// Will create a RoleBinding in the same namespace as the managedClusterAddon to bind the user
+// provided ClusterRole/Role to the "system:open-cluster-management:cluster:<cluster-name>:addon:<addon-name>"
+// Group.
+type HubPermissionConfig struct {
+	// Type of the permissions setting. It defines how to bind the roleRef
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=CurrentCluster;SingleNamespace
+	Type HubPermissionsBindingType `json:"type"`
+
+	// RoleRef is an reference to the permission resource. it could be a role or a cluster role,
+	// the user must make sure it exist on the hub cluster.
+	// +kubebuilder:validation:Required
+	RoleRef rbacv1.RoleRef `json:"roleRef"`
```

Comment:
A cluster role ref may be convenient here.  We would expect this to be the same in every namespace, rigth?